### PR TITLE
`Development`: Restructure non-notification mails to use data transfer objects

### PR DESCRIPTION
**Pull request by xHadie**
#### General

- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context

The usage of DTOs prevents the leakage of potentially sensitive information, makes use cases clearer, and results in more efficient data transfer. 
In the context of the mail service, it allows developers to easily see what information is actually being transferred in a mail. 

Furthermore, this PR introduces a straightforward testing strategy for HTML mail content generated by thymeleaf. 
This ensures that the shift to DTOs within this PR did not break the interpretation and send logic of the mail messages. In addition, it allows us to test that DTO content is really being used within the mail content, making editing the scopes of DTOs and sending methods more convenient.

During development, it revealed, for instance, that the `dataExportFailedAdminEmail.html` template used the recipient username instead of the username connected to the data export.

In case this PR gets accepted, we can apply the same approach to the notification-based mails and also refactor the underlying bulk of thymeleaf fragments.

Implements a part of #9524

### Description

**Architecture:**
- Implemented `AbstractMailContentTest` which allows testing HTML content generated by thymeleaf via `getGeneratedEmailTemplateText` for the correct rendering of text
- Implemented `IMailRecipientUserDTO` which contains the minimal set of required attributes a mail receiver needs to have to be compatible with `MailSendingService`. Individual mail templates will implement this interface in a dedicated recipient record, which may be extended with additional attributes
- Adjusted `MailSendingService::sendMail` to use `IMailRecipientUserDTO` instead of `User`
- Adjusted `MailService::createBaseContext`, `MailService::sendEmailFromTemplate`, `MailService::prepareTemplateAndSendEmail` and `MailService::prepareTemplateAndSendEmailWithArgumentInSubject` to use `IMailRecipientUserDTO` instead of `User`

**[Use case] Activation mail:**
- Added `ActivationMailRecipientDTO` and the accompanying `ActivationMailTest` and adjusted `MailService::sendActivationEmail` to use the DTO
- Adjusted `activationEmail.html`

**[Use  case] Password reset mail:**
- same as activation mail

**[Use  case] SAML2 password reset mail:**
- same as activation mail

**[Use case] Data export failed mail:**
- Added `DataExportFailedContentDTO` which contains only information that is relevant to the template
- Added `DataExportFailedMailAdminRecipientDTO` and the accompanying `ActivationMailTest` and adjusted `MailService::sendActivationEmail` to use the DTOs
- Adjusted the parameter list of `MailService::sendDataExportFailedEmailForAdmin`
- Adjusted `dataExportFailedAdminEmail.html`

**[Use case] Data export successful mail:**
- Added `DataExportSuccessfulContentsDTO` and `DataExportSuccessfulContentDTO` which contain only information that is relevant to the template.
- Added `DataExportSuccessfulMailAdminRecipientDTO` and the accompanying `DataExportSuccessfulMailTest` and adjusted `MailService::sendSuccessfulDataExportsEmailToAdmin` to use the dto
- Adjusted the parameter list of `MailService::sendSuccessfulDataExportsEmailToAdmin`
- Adjusted `successfulDataExportsAdminEmail.html`

**[Use  case] Notifications mails:**
- Added `NotificationMailRecipientDTO` and adjusted `MailService::sendNotification`
- This is a minimal required change. Notifications will be adjusted to use DTOs in a separate PR 

**Additional Improvements:**
- Reduced the visibility of `MailService::sendSuccessfulDataExportsEmailToAdmin` from `public` to `private` and adjusted `DataExportScheduleServiceTest` to use the parent method with an equal name
- Reduced the visibility of `MailService::sendDataExportFailedEmailForAdmin` from `public` to `private`
- Reduced the visibility of `MailService::sendEmailFromTemplate` from `public` to `private`
- Fixed wrong username usage in `dataExportFailedAdminEmail.html`

**Identified issues:**
- `creationEmail.html` is not being used
- Missing structures and naming conventions for mail templates and in general mail-related project substructures
- `MailService` needs refactoring and is doing too much. Some of the private methods are relevant only for single-use cases. In my opinion, it would make sense to extract individual mail construction into their own classes and let `MailService` delegate the constructed mails to `MailSendingService`.

### Steps for Testing [To be extended after code reviews]

Prerequisites:
- 1 Admin
- 1 Students

**FOREACH IN** [`Activation mail`, `Password reset mail`, `SAML2 set password mail`, `Data export failed mail`, `Data export successful mail`, `Weekly summary mail`, `Some notification mail`]  **DO**:
1. [...] do steps to produce the mail
2. Verify that the content in the email in german and english s correct

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

| Class/File                                     | Line Coverage | Confirmation (assert/expect) |
| ---------------------------------------------- | ------------: | ---------------------------: |
| ActivationMailRecipientDTO.java                |          100% |                            ✅ |
| DataExportFailedContentDTO.java                |          100% |                            ✅ |
| DataExportSuccessfulContentDTO.java            |          100% |                            ✅ |
| DataExportSuccessfulContentsDTO.java           |          100% |                            ✅ |
| DataExportSuccessfulMailAdminRecipientDTO.java |          100% |                            ✅ |
| NotificationMailRecipientDTO.java              |             0 |                            ❌ |
| PasswordResetRecipientDTO.java                 |          100% |                            ✅ |
| SAML2SetPasswordMailRecipientDTO.java          |          100% |                            ✅ |
| WeeklyExerciseSummaryDTO.java                  |          100% |                            ✅ |
| WeeklySummaryMailContentDTO.java               |          100% |                            ✅ |
| WeeklySummaryMailRecipientDTO.java             |          100% |                            ✅ |
| IMailRecipientUserDTO.java                     |          100% |                            ✅ |
| MailSendingService.java                        |          100% |                            ✅ |
| MailService.java                               |           50% |                            ✅ |

## Summary by CodeRabbit

- **New Features**
	- Introduced new Data Transfer Objects (DTOs) for various email notifications, enhancing email recipient handling.
	- Updated email templates to reflect new data structures and improve clarity in notifications.

- **Bug Fixes**
	- Corrected variable references in email templates for better data accuracy.

- **Tests**
	- Added new test classes to verify email content generation for different scenarios, ensuring correct data injection into templates.

### CODE REVIEW

File: src/main/java/de/tum/cit/aet/artemis/communication/service/notifications/MailService.java [hunk 1]
```diff
@@ -224,6 +239,7 @@ public void sendNotification(Notification notification, Set<User> users, Object
      * @param user                who should be contacted
      * @param notificationSubject that is used to provide further information (e.g. exercise, attachment, post, etc.)
      */
+
```
Comment (BBesrour) [inline src/main/java/de/tum/cit/aet/artemis/communication/service/notifications/MailService.java:242]: Remove this new line

**Review by BBesrour — APPROVED** [review]
code lgtm. nice changes! Just a quick reminder to open the PR (out of draft mode) and add the "Ready for Review" label when it's ready.

File: src/main/java/de/tum/cit/aet/artemis/communication/service/notifications/MailService.java (hunk 1 above)
Comment (xHadie) [reply src/main/java/de/tum/cit/aet/artemis/communication/service/notifications/MailService.java]: Thank you for the review and the heads up!
I fixed the server-style issues and removed the new line.

**Review by MarkusPaulsen — CHANGES_REQUESTED** [review]
Added my review

**Pull request closed without merging**